File: SGUEES-API/sguees.api/Options/SelectionHiring/SC_FRECUENCIA/SC_FRECUENCIAService.cs
// Qué hace: lógica de negocio del catálogo frecuencia.
// Cómo: valida los datos y llama a ISC_FRECUENCIARepository para ejecutar el CRUD.
using System.Collections.Generic;
using System.Threading.Tasks;
using eFramework.Core;
using SGUEES.Models;
using SGUEES.Repositories;

namespace SGUEES.Services
{
    // Qué hace: servicio de frecuencia.
    // Cómo: valida los datos y llama al repositorio para persistir la información.
    public class SC_FRECUENCIAService : ISC_FRECUENCIAService
    {
        private readonly ISC_FRECUENCIARepository _repo;

        public SC_FRECUENCIAService(ISC_FRECUENCIARepository repo)
        {
            _repo = repo;
        }

        // Qué hace: obtiene el listado de frecuencias.
        // Cómo: llama a GetAllAsync del repositorio con los parámetros armados por BuildParameters.
        public async Task<CResult> GetAllAsync(SC_FRECUENCIAParam xWhere)
        {
            return await _repo.GetAllAsync(BuildParameters(xWhere));
        }

        // Qué hace: obtiene solo las frecuencias activas de la empresa.
        // Cómo: llama a GetFrecuenciasActivasAsync del repositorio filtrando por CORR_EMPRESA.
        public async Task<CResult> GetFrecuenciasActivasAsync(SC_FRECUENCIAParam xWhere)
        {
            var p = new List<CParameter>
            {
                new CParameter() { ParameterName = "CORR_EMPRESA", Value = xWhere.CORR_EMPRESA, DbType = System.Data.DbType.Int32 },
            };

            return await _repo.GetFrecuenciasActivasAsync(p);
        }

        // Qué hace: obtiene una frecuencia puntual.
        // Cómo: llama a GetAsync del repositorio filtrando por CORR_EMPRESA y CORR_FRECUENCIA.
        public async Task<CResult> GetAsync(SC_FRECUENCIAParam xWhere)
        {
            var p = new List<CParameter>
            {
                new CParameter() { ParameterName = "CORR_EMPRESA", Value = xWhere.CORR_EMPRESA, DbType = System.Data.DbType.Int32 },
                new CParameter() { ParameterName = "CORR_FRECUENCIA", Value = xWhere.CORR_FRECUENCIA, DbType = System.Data.DbType.Int32 },
            };

            return await _repo.GetAsync(p);
        }

        // Qué hace: crea una frecuencia.
        // Cómo: valida la empresa de sesión y los datos con ValidateEmpresaSesion y Validate, normaliza con NormalizeData y llama a CreateAsync del repositorio.
        public async Task<CResult> CreateAsync(SC_FRECUENCIATable Data, string vLOGIN_SISTEMA, string vESTACION)
        {
            var empresaError = ValidateEmpresaSesion(Data.CORR_EMPRESA);
            if (empresaError != null)
            {
                return empresaError;
            }

            var validation = Validate(Data);
            if (validation != null)
            {
                return validation;
            }

            NormalizeData(Data);
            return await _repo.CreateAsync(Data, vLOGIN_SISTEMA, vESTACION);
        }

        // Qué hace: actualiza una frecuencia.
        // Cómo: valida la empresa de sesión, los datos y el CORR_FRECUENCIA a actualizar, normaliza con NormalizeData y llama a UpdateAsync del repositorio.
        public async Task<CResult> UpdateAsync(SC_FRECUENCIATable Data, string vLOGIN_SISTEMA, string vESTACION)
        {
            var empresaError = ValidateEmpresaSesion(Data.CORR_EMPRESA);
            if (empresaError != null)
            {
                return empresaError;
            }

            var validation = Validate(Data);
            if (validation != null)
            {
                return validation;
            }

            if (Data.CORR_FRECUENCIA <= 0)
            {
                return ValidationError("No se pudo identificar la frecuencia a actualizar.");
            }

            NormalizeData(Data);
            return await _repo.UpdateAsync(Data, vLOGIN_SISTEMA, vESTACION);
        }

        // Qué hace: elimina una frecuencia.
        // Cómo: valida la empresa de sesión con ValidateEmpresaSesion y llama a DeleteAsync del repositorio.
        public async Task<CResult> DeleteAsync(SC_FRECUENCIATable Data, string vLOGIN_SISTEMA, string vESTACION)
        {
            var empresaError = ValidateEmpresaSesion(Data.CORR_EMPRESA);
            if (empresaError != null)
            {
                return empresaError;
            }

            return await _repo.DeleteAsync(Data, vLOGIN_SISTEMA, vESTACION);
        }

        // Qué hace: cambia el estado activo/inactivo de una frecuencia.
        // Cómo: valida la empresa de sesión y el CORR_FRECUENCIA a actualizar, y llama a ActivarInactivarAsync del repositorio.
        public async Task<CResult> ActivarInactivarAsync(SC_FRECUENCIATable Data, string vLOGIN_SISTEMA, string vESTACION)
        {
            var empresaError = ValidateEmpresaSesion(Data.CORR_EMPRESA);
            if (empresaError != null)
            {
                return empresaError;
            }

            if (Data.CORR_FRECUENCIA <= 0)
            {
                return ValidationError("No se pudo identificar la frecuencia a actualizar.");
            }

            return await _repo.ActivarInactivarAsync(Data, vLOGIN_SISTEMA, vESTACION);
        }

        // Qué hace: arma los parámetros de filtro para el repositorio.
        // Cómo: construye la lista con CORR_EMPRESA a partir de xWhere.
        private static List<CParameter> BuildParameters(SC_FRECUENCIAParam xWhere)
        {
            return new List<CParameter>
            {
                new CParameter() { ParameterName = "CORR_EMPRESA", Value = xWhere.CORR_EMPRESA, DbType = System.Data.DbType.Int32 },
            };
        }

        // Qué hace: normaliza los datos antes de guardar.
        // Cómo: recorta espacios de NOMBRE_FRECUENCIA y aplica ESTADO_FRECUENCIA activo cuando no viene informado.
        private static void NormalizeData(SC_FRECUENCIATable Data)
        {
            Data.NOMBRE_FRECUENCIA = Data.NOMBRE_FRECUENCIA?.Trim();
            Data.ESTADO_FRECUENCIA ??= true;
        }

        // Qué hace: valida los datos obligatorios de la frecuencia.
        // Cómo: comprueba que Data no sea nulo y que NOMBRE_FRECUENCIA no esté vacío ni supere 50 caracteres.
        private static CResult Validate(SC_FRECUENCIATable Data)
        {
            if (Data == null)
            {
                return ValidationError("No se recibieron datos de la frecuencia.");
            }

            if (string.IsNullOrWhiteSpace(Data.NOMBRE_FRECUENCIA))
            {
                return ValidationError("Debe ingresar el nombre de la frecuencia.");
            }

            if (Data.NOMBRE_FRECUENCIA.Trim().Length > 50)
            {
                return ValidationError("El nombre de la frecuencia no puede superar 50 caracteres.");
            }

            return null;
        }

        // Qué hace: verifica que exista empresa en la sesión.
        // Cómo: si corrEmpresa es mayor a cero devuelve null; de lo contrario devuelve un CResult con el error de empresa no asignada.
        private static CResult ValidateEmpresaSesion(int corrEmpresa)
        {
            if (corrEmpresa > 0)
            {
                return null;
            }

            return new CResult
            {
                Data = null,
                Result = false,
                CodeHelper = 0,
                ErrorCode = 4100,
                ErrorMessage = "No se pudo guardar la frecuencia porque su usuario no tiene una empresa asignada. Solicite que le configuren una empresa por defecto en el sistema.",
                ErrorSource = "[SC_FRECUENCIAService]",
                RowsAffected = 0
            };
        }

        // Qué hace: construye un resultado de error de validación.
        // Cómo: arma un CResult con Result en false y el mensaje recibido.
        private static CResult ValidationError(string message)
        {
            return new CResult
            {
                Data = null,
                Result = false,
                CodeHelper = 0,
                ErrorCode = -1,
                ErrorMessage = message,
                ErrorSource = "[SC_FRECUENCIAService]",
                RowsAffected = 0
            };
        }
    }
}

File: SGUEES-SPA/src/app/pages/SelectionHiring/sc-frecuencia/sc-frecuencia.service.ts
// Qué hace: servicio de negocio del catálogo Frecuencia.
// Cómo: valida los datos, ejecuta el CRUD a través del repositorio y arma la configuración de grilla y formulario.
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { IParam } from 'src/app/FxAPI/IParam';
import { IResult } from 'src/app/FxAPI/IResult';
import { NotifyType } from 'src/app/shared/models/NotifyType';
import { buildAuditGridColumns } from 'src/app/shared/mtto/mtto-grid.helpers';
import { createEstadoColumnConfig, ESTADO_ACTIVO_INACTIVO_LABELS } from 'src/app/shared/utils/remote-grid-filter.util';
import { ScFrecuencia } from './models/sc-frecuencia';
import { ScFrecuenciaRepository } from './sc-frecuencia.repository';

const ESTADO_FIELD = 'ESTADO_FRECUENCIA';

@Injectable({ providedIn: 'root' })
// Qué hace: servicio de frecuencia.
// Cómo: valida los datos y llama a ScFrecuenciaRepository para ejecutar el CRUD.
export class ScFrecuenciaService {
	constructor(private repo: ScFrecuenciaRepository) {}

	// Qué hace: valida el formulario de frecuencia antes de guardar.
	// Cómo: revisa que NOMBRE_FRECUENCIA no esté vacío y no supere 50 caracteres, notificando con msg cuando falla.
	esValido(model: ScFrecuencia, msg: Function): boolean {
		if (!model.NOMBRE_FRECUENCIA || model.NOMBRE_FRECUENCIA.trim() === '') {
			msg('Debe ingresar el nombre de la frecuencia.', NotifyType.Warning);
			return false;
		}

		if (model.NOMBRE_FRECUENCIA.trim().length > 50) {
			msg('El nombre de la frecuencia no puede superar 50 caracteres.', NotifyType.Warning);
			return false;
		}

		return true;
	}

	// Qué hace: obtiene el listado de frecuencias.
	// Cómo: llama a getAll del repositorio con el filtro construido por buildWhere.
	getAll(param: any): Observable<IResult> {
		return this.repo.getAll(this.buildWhere(param));
	}

	// Qué hace: obtiene una frecuencia puntual.
	// Cómo: llama a get del repositorio filtrando por CORR_FRECUENCIA.
	get(param: any): Observable<IResult> {
		return this.repo.get([{ Parameter: 'CORR_FRECUENCIA', Value: param.CORR_FRECUENCIA }]);
	}

	// Qué hace: crea una nueva frecuencia.
	// Cómo: llama a create del repositorio con el modelo recibido.
	insert(model: any): Observable<IResult> {
		return this.repo.create(model);
	}

	// Qué hace: actualiza una frecuencia existente.
	// Cómo: llama a update del repositorio con el modelo y su CORR_FRECUENCIA.
	update(model: any): Observable<IResult> {
		return this.repo.update(model, [{ Parameter: 'CORR_FRECUENCIA', Value: model.CORR_FRECUENCIA }]);
	}

	// Qué hace: elimina una frecuencia.
	// Cómo: llama a delete del repositorio filtrando por CORR_FRECUENCIA.
	delete(model: any): Observable<IResult> {
		return this.repo.delete([{ Parameter: 'CORR_FRECUENCIA', Value: model.CORR_FRECUENCIA }]);
	}

	// Qué hace: cambia el estado activo/inactivo de una frecuencia.
	// Cómo: llama a activarInactivar del repositorio filtrando por CORR_FRECUENCIA.
	activarInactivar(model: any): Observable<IResult> {
		return this.repo.activarInactivar(model, [{ Parameter: 'CORR_FRECUENCIA', Value: model.CORR_FRECUENCIA }]);
	}

	// Qué hace: define columnas y formatos de la grilla de mantenimiento.
	// Cómo: arma el arreglo de columnas (correlativo, nombre, estado y auditoría) usado por app-data-grid-mtto.
	getColumns(): any {
		return [
			{
				dataField: 'CORR_FRECUENCIA',
				caption: 'Corr.',
				width: 90,
				dataType: 'number',
				filterOperations: ['=', '<', '>', '<=', '>='],
			},
			{ dataField: 'NOMBRE_FRECUENCIA', caption: 'Frecuencia', width: 300 },
			createEstadoColumnConfig(ESTADO_FIELD, ESTADO_ACTIVO_INACTIVO_LABELS),
			...buildAuditGridColumns({ withDateTimeFilter: true }),
		];
	}

	// Qué hace: configura el contador de registros de la grilla.
	// Cómo: define el resumen totalItems que cuenta CORR_FRECUENCIA.
	getSummary(): any {
		return {
			totalItems: [
				{
					column: 'CORR_FRECUENCIA',
					summaryType: 'count',
					valueFormat: '#,##0',
					displayFormat: 'Cant: {0}',
				},
			],
		};
	}

	// Qué hace: define los campos y reglas del formulario de frecuencia.
	// Cómo: arma el arreglo de items (correlativo, nombre y estado) usado por dx-form.
	getItems(): any {
		return [
			{ dataField: 'CORR_FRECUENCIA', label: { text: 'Corr.' }, colSpan: 1, editorOptions: { readOnly: true } },
			{
				dataField: 'NOMBRE_FRECUENCIA',
				label: { text: 'Nombre frecuencia' },
				colSpan: 5,
				editorOptions: { placeholder: 'Nombre frecuencia...', showClearButton: true, maxLength: 50 },
				validationRules: [{ type: 'required', message: 'Este campo es obligatorio' }],
			},
			{ dataField: 'ESTADO_FRECUENCIA', label: { text: 'Activo' }, editorType: 'dxCheckBox', colSpan: 2 },
		];
	}

	// Qué hace: traduce los filtros del componente al formato esperado por la API.
	// Cómo: agrega a xWhere el parámetro CORR_FRECUENCIA cuando viene informado en param.
	private buildWhere(param: any): IParam[] {
		const xWhere: IParam[] = [];

		if (param.CORR_FRECUENCIA) {
			xWhere.push({ Parameter: 'CORR_FRECUENCIA', Value: param.CORR_FRECUENCIA });
		}

		return xWhere;
	}
}
